File: src/core/tools/builtin/websearch.ts
/**
 * WebSearch Tool - Search the web for current information
 */

import { z } from 'zod';
import type { Tool, ToolContext, ToolResult } from '../types.js';
import { getErrorMessage } from '../types.js';
import {
  createSearchProvider,
  getCurrentSearchProviderName,
  type SearchResult,
} from '../../providers/search/index.js';
import { loadToolDescription } from '../../../cli/prompts/index.js';

// Constants
const DEFAULT_NUM_RESULTS = 10;

// Input schema
export const WebSearchInputSchema = z.object({
  query: z
    .string()
    .min(2)
    .describe('The search query (minimum 2 characters)'),
  allowed_domains: z
    .array(z.string())
    .optional()
    .describe('Only include results from these domains'),
  blocked_domains: z
    .array(z.string())
    .optional()
    .describe('Exclude results from these domains'),
  num_results: z
    .number()
    .optional()
    .describe(`Number of results to return (default: ${DEFAULT_NUM_RESULTS})`),
});
export type WebSearchInput = z.infer<typeof WebSearchInputSchema>;

/**
 * Format search results as markdown
 */
function formatResults(results: SearchResult[], query: string): string {
  if (results.length === 0) {
    return `No results found for "${query}".`;
  }

  const lines: string[] = [`Found ${results.length} results for "${query}":\n`];

  results.forEach((result, index) => {
    lines.push(`${index + 1}. [${result.title}](${result.url})`);
    if (result.snippet) {
      lines.push(`   ${result.snippet}\n`);
    } else {
      lines.push('');
    }
  });

  return lines.join('\n');
}

/**
 * WebSearch Tool
 */
export const websearchTool: Tool<WebSearchInput> = {
  name: 'WebSearch',
  description: loadToolDescription('websearch'),
  parameters: WebSearchInputSchema,

  async execute(input: WebSearchInput, context: ToolContext): Promise<ToolResult> {
    const startTime = Date.now();

    try {
      const provider = createSearchProvider();

      const results = await provider.search(input.query, {
        numResults: input.num_results ?? DEFAULT_NUM_RESULTS,
        allowedDomains: input.allowed_domains,
        blockedDomains: input.blocked_domains,
        abortSignal: context.abortSignal,
      });

      const output = formatResults(results, input.query);
      const duration = Date.now() - startTime;
      const providerName = getCurrentSearchProviderName();

      return {
        success: true,
        output,
        metadata: {
          title: `Search("${input.query}")`,
          subtitle: `Found ${results.length} results via ${providerName}`,
          duration,
        },
      };
    } catch (error) {
      return {
        success: false,
        output: '',
        error: `Search failed: ${getErrorMessage(error)}`,
      };
    }
  },
};

File: internal/tool/web/websearch.go
package web

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/yanmxa/gencode/internal/llm/search"
	"github.com/yanmxa/gencode/internal/tool"
	"github.com/yanmxa/gencode/internal/tool/toolresult"
)

// WebSearchTool searches the web for information.
// SearchProviderGetter returns the configured search provider name at runtime.
// If nil or returns empty, the default provider is used.
type WebSearchTool struct {
	SearchProviderGetter func() string
}

func (t *WebSearchTool) Name() string        { return "WebSearch" }
func (t *WebSearchTool) Description() string { return "Search the web for up-to-date information" }
func (t *WebSearchTool) Icon() string        { return toolresult.IconWeb }

func (t *WebSearchTool) Execute(ctx context.Context, params map[string]any, cwd string) toolresult.ToolResult {
	start := time.Now()

	query, err := tool.RequireString(params, "query")
	if err != nil {
		return toolresult.NewErrorResult(t.Name(), err.Error())
	}

	numResults := tool.GetInt(params, "num_results", 10)

	// Get optional domain filters
	var allowedDomains, blockedDomains []string
	if domains, ok := params["allowed_domains"].([]any); ok {
		for _, d := range domains {
			if s, ok := d.(string); ok {
				allowedDomains = append(allowedDomains, s)
			}
		}
	}
	if domains, ok := params["blocked_domains"].([]any); ok {
		for _, d := range domains {
			if s, ok := d.(string); ok {
				blockedDomains = append(blockedDomains, s)
			}
		}
	}

	// Get the configured search provider
	var searchProvider search.Provider
	if t.SearchProviderGetter != nil {
		if name := t.SearchProviderGetter(); name != "" {
			searchProvider = search.CreateProvider(search.ProviderName(name))
		}
	}
	if searchProvider == nil {
		searchProvider = search.GetDefaultProvider()
	}

	// Execute search
	opts := search.SearchOptions{
		NumResults:     numResults,
		AllowedDomains: allowedDomains,
		BlockedDomains: blockedDomains,
		Timeout:        30 * time.Second,
	}

	results, err := searchProvider.Search(ctx, query, opts)
	if err != nil {
		return toolresult.NewErrorResult(t.Name(), fmt.Sprintf("search failed: %v", err))
	}

	// Format results as Markdown
	var sb strings.Builder
	if len(results) == 0 {
		sb.WriteString("No results found for: " + query)
	} else {
		sb.WriteString(fmt.Sprintf("Found %d results for: %s\n\n", len(results), query))
		for _, r := range results {
			sb.WriteString(fmt.Sprintf("- [%s](%s)\n", r.Title, r.URL))
			if r.Snippet != "" {
				sb.WriteString(fmt.Sprintf("  %s\n\n", r.Snippet))
			}
		}
	}

	duration := time.Since(start)

	return toolresult.ToolResult{
		Success: true,
		Output:  sb.String(),
		Metadata: toolresult.ResultMetadata{
			Title:     t.Name(),
			Icon:      t.Icon(),
			Subtitle:  fmt.Sprintf("%s via %s", query, searchProvider.DisplayName()),
			ItemCount: len(results),
			Duration:  duration,
		},
	}
}

// SetSearchProviderGetter configures the provider name getter on the registered
// WebSearchTool instance. Call this from the app layer after the provider store
// is available.
func SetSearchProviderGetter(getter func() string) {
	if t, ok := tool.Get("WebSearch"); ok {
		if wst, ok := t.(*WebSearchTool); ok {
			wst.SearchProviderGetter = getter
		}
	}
}

func init() {
	tool.Register(&WebSearchTool{})
}
